lister: add directory_lister

bhalevy opened the pull request:
directory_lister provides a simpler interface compared to lister.

After creating the directory_lister,
its async get() method should be called repeatedly,
returning a std::optional<directory_entry> each call,
until it returns a disengaged entry or an error.
    
This is especially suitable for coroutines
as demonstrated in the unit tests that were added.
    
For example:
```c++
        auto dl = directory_lister(path);
        while (auto de = co_await dl.get()) {
            co_await process(*de);
        }
```

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy commented:
Cc @StarostaGit

nyh commented on lister.cc:
```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
+        auto f = co_await open_checked_directory(general_disk_error_handler, _dir.native());
+        auto walker = [this] (fs::path dir, directory_entry de) {
+            return _q.push_eventually(std::make_optional<directory_entry>(std::move(de))).handle_exception([] (std::exception_ptr) {
+                // ignore exception, queue was aborted
```
If I understand correctly, this means that if the queue is aborted, the lister will continue to list potentially a million more files, and ignore them one by one. Wouldn't it be better to pass the exception through, and presumably the lister will stop listing?

bhalevy replied:
It might be possible, but requires adding support for properly aborting the lister.

nyh replied:
> It might be possible, but requires adding support for properly aborting the lister.

I was assuming (without checking the code) that if the walker throws an exception, the lister will properly abort. Doesn't it?
Your code deliberately catches the exception that the walker throws. I was thinking, maybe if it did throw - everything will work as expected?

bhalevy replied:
it did throw and all hell broke loose :)
I'm exaggerating a bit, but I wish it worked as simple as returning an exception from the walker.

nyh replied:
So maybe you discovered a bug in the existing lister, which should be fixed?

The user-written "walker" is supposed to return a future, and this future **may** be exceptional, and the lister needs to support this case in a reasonable way - and if it doesn't, I would think it's a bug. 

NOTE: I'm not suggesting that walker should throw (it is probably called without futurize() protection), but rather that it can return an exceptional future. I think this is what happens in coroutines automatically.

bhalevy replied:
Fixed in v2

nyh commented on lister.cc:
```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
+                // ignore exception, queue was aborted
+            });
+        };
+        assert(!_lister);
```
Nitpick: the `if(!_done)` above was weird to read (it sounds like "not done", but it's not what being tested). And here you have an `assert(!_lister)`. Wouldn't the code be easier to understand if instead you had instead of `if (!_done)` a  `if (!_lister)`?

bhalevy replied:
It was `if (!_lister)` in an earlier message, but I want `close()` to get rid of `_lister` and then if `get()` is called after `close()` I don't want it to re-create the `_lister`.  I could add a `bool _closed` member to prevent, but instead I overloaded the `_done` future that starts disengaged and is engaged once when the `_lister` is first created, and possibly on `close()` it is set to a ready future, but remains engaged.  Maybe it could be named better so it won't look weird.

nyh replied:
I see. I actually understood that at the end of my review but wrote this comment in the middle.
So ok. The code seems correct, I guess there are seven other ways to write it, some will look more natural to me, but nevermind.

nyh commented on lister.cc:
```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
+        };
+        assert(!_lister);
+        _lister = std::make_unique<lister>(std::move(f), _type, std::move(walker), _filter, _dir, _do_show_hidden);
+        auto done_fut = _lister->done().then_wrapped([this] (future<> f) {
```
nitpick: I think you can just do `_done` = ... here. I think (but didn't test) that when _done is optional<future> you can assign to it a future and that's it - I think you don't need to call `std::make_optional` as you did below.

bhalevy replied:
I'll try that...

bhalevy replied:
Fixed in v3

nyh commented on lister.cc:
```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
+        });
+        _done = std::make_optional<future<>>(std::move(done_fut));
+    }
+    co_return co_await _q.pop_eventually();
```
Unless I missed something you're not using coroutines here for anything else, so you can do more simply (and I think more efficiently?)
`return _q.pop_eventaully();`

bhalevy replied:
I call `co_await open_checked_directory` above...

nyh replied:
So I did miss something :-)

nyh commented on lister.hh:
```diff
@@ -166,10 +167,54 @@ private:
+    std::unique_ptr<lister> _lister;
+    std::optional<future<>> _done;
+
+    using queue_entry_type = std::optional<directory_entry>;
```
Why is the queue type an optional<directory_entry> and not just directory_entry?
I understand that get() will return an optional<directory_entry>, but not why every item in the queue needs to be an optional. It looks like a mistake?

bhalevy replied:
It's for pushing an in-band "end of queue" entry.
seastar::queue doesn't have an out of band "end-of-stream" concept IIUC

nyh replied:
Ok, somehow I missed that. pipe (which I told you is very similar to what you did :-)) is also using the same trick. I forgot it's part of pipe, not of queue.

nyh commented on lister.hh:
```diff
@@ -166,10 +167,54 @@ private:
+
+static inline fs::path operator/(const fs::path& lhs, std::string_view rhs) {
+    return lhs / fs::path(rhs);
+}
```
Nitpick: I wonder if it's not enough to have this std::string_view version, and it will work for `const sstring&` and `const std::string&` because of implicit conversion. But maybe not, I didn't check.

bhalevy replied:
I think it works for seastar::sstring, but unfortunately it doesn't work for std::string

nyh replied:
Ok, I wonder why, but since you tested, ok, I'll resolve this discussion.

nyh commented on test/boost/lister_test.cc:
```diff
@@ -0,0 +1,183 @@
+    auto ql = queue_lister(tmp.path());
+
+    while (auto de = co_await ql.get()) {
+        assert(de->type);
```
I don't think we should be using assert() in boost-tests. I think Boost has its own idioms for checks. Of course if the test doesn't fail, it doesn't matter... But if it does fail, we'll get a core dump instead of a normal boost test report, right?

bhalevy replied:
I didn't use `BOOST_REQUIRE` so not to spam the log with hundreds of boost messages.
If boost had a check that logs nothing by default I'd use it happily.

nyh replied:
using assert in tests works great in pytest, but in Boost tests, if this test ever fails, it will cause a crash instead of normal error messages. That just doesn't sound right.
I don't know about the spamming. Why should Boost log ever **successful** check? I don't remember that it does. Please re-check, and perhaps consult with people who are bigger fans than me with that test framework (I am *not* a fan).

bhalevy replied:
with `-l all` it prints successes as well. It's not the default though, but I use it in my scripts.

nyh replied:
So you use a non-default option to make tests more spammy, and then surprised when it's too spammy? :-)

bhalevy replied:
Fixed in v3

nyh commented on test/boost/lister_test.cc:
```diff
@@ -0,0 +1,183 @@
+            assert(de);
+        }
+        BOOST_TEST_MESSAGE("Closing queue_lister");
+        co_await ql.close();
```
Unless I'm missing something, this block is an **exact** copy of the previous block - so why do it twice? You can do the extra `ql.get()` check that you do below, after the first block - no need for the second block at all.

bhalevy replied:
the first one destroys the `queue_lister` after `close()`,
this one, after the extra `get()` call.

Maybe I can extract the inner-body of the block into a worker function...

nyh replied:
But who cares about this difference? the extra get() call just fails and you check that - and then you can destroy. Why do you need both with and without the get? I don't see what it checks additionally.

bhalevy replied:
I want to make sure that the queue_lister can be destroyed safely right after close, possibly with stuff in the queue, vs. destroyed after `get()` returns an error.  It needs to ensure that the underlying open directory is closed properly (and by the way, it's not when the walker throws an exception IIRC - needs fixing)

bhalevy commented on lister.cc:
```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
     });
 }
+
+future<std::optional<directory_entry>> queue_lister::get() {
```
I need to test that.
Doesn't sound too bad :)

bhalevy replied:
pipe didn't work well for me.
it isn't really abortable so it's a nightmare stopping in the middle of the directory listing.
you always need to consume all entries.

nyh replied:
What makes it not abortable? Could we fix pipe - or something like it - to be abortable?
Anyway, this was just an idea. It's not a must.

bhalevy replied:
> What makes it not abortable? Could we fix pipe - or something like it - to be abortable?
> Anyway, this was just an idea. It's not a must.

There's no function to close either the writer or the reader. Destroying the pipe while there's a blocking reader is fragile since it is synchronous and doesn't wait for the read to actually pop the "eof" marker.

I think this could be potentially fixed, but out of scope.
Just using the queue is simple enough and I don't see what value a pipe brings here.

nyh replied:
> There's no function to close either the writer or the reader. Destroying the pipe while there's a blocking reader is fragile since it is synchronous and doesn't wait for the read to actually pop the "eof" marker.

This was a design goal of pipes - that you have two **uncoupled** ends, the reader and writer, that can destroy their ends separately (there is no overall "pipe" object that you need to worry about). When the writer destroys his end, he can be assured that the reader will be woken up and read the eof - but can't wait for this to happen and doesn't want to wait for it to happen. It's usually completely pointless to wait for the reader to receive the eof - because, why wait to receive the eof, and not wait for - say - the reader to **handle** the eof?

So this decoupled ends is a feature of pipes, not something we'll want to fix. Maybe it's something we should even learn from: Yes, it's easy and sounds very responsible to have a `close()` method on everything and force the users to use it, but sometimes, we can invent simpler APIs that don't need an explicit `close()`.

Anyway, never mind, we don't have to use pipes :-)

bhalevy commented on lister.hh:
```diff
@@ -166,10 +167,54 @@ private:
     future<directory_entry> guarantee_type(directory_entry de);
 };
 
+class queue_lister {
```
I'm not married to that name.
Maybe you have a better idea for it?

bhalevy replied:
Fixed in v2

bhalevy commented:
In v2:
- Renamed `queue_lister` to `directory_lister`
- Simplified the implementation
- Used directory_lister in `sstable_directory::process_sstable_dir`
  - while converting it to a coroutine

bhalevy commented:
In v3:
- Assign to _done directly (no need for a `done_fut` temporary)
- Converted the business-logic related asserts to BOOST_REQUIRE in unit test
  - In contrast to some test internals that are still asserted, i.e. their failure means something wrong in the test, not in the system.

nyh reviewed (changes requested):
Looks good to me. I just left a few minor comments/requests.

bhalevy commented on lister.hh:
```diff
@@ -153,6 +154,41 @@ private:
+    void abort(std::exception_ptr ex);
+
+    // Close the directory_lister, ignoring any errors.
+    // Must be called if get() did not drain the queue.
```
That's correct.

`close()` never returns an error.

The main job of `close()` is to wait for the underlying lister to be `done()`,
anything else is a bonus.

This is needed, as I wrote in the comment, in case 'get()` did not drain the queue, until it got a disengaged `std::optional<directory_entry>`.  If it did then `close()` is not required.
The motivation to have it is when one lists a directory until some specific entry is found, then close() can be used to skip over anything else beyond this entry, without having to wait for all entries to be listed.

The call to `seastar::queue::abort()` will indeed abort any waiting `get()`.

bhalevy commented on lister.cc:
```diff
@@ -75,3 +77,50 @@ future<> lister::rmdir(fs::path dir) {
+}
+
+future<std::optional<directory_entry>> directory_lister::get() {
+    if (!_done) {
```
The boolean check is since it's an optional future.
We can call it `_opt_done_fut`.

bhalevy replied:
Improved documentation in v4.

bhalevy commented on lister.cc:
```diff
@@ -75,3 +77,50 @@ future<> lister::rmdir(fs::path dir) {
+
+future<std::optional<directory_entry>> directory_lister::get() {
+    if (!_done) {
+        auto f = co_await open_checked_directory(general_disk_error_handler, _dir.native());
```
ok, I think `dir` is appropriate here.

bhalevy replied:
Renamed in v4.

bhalevy commented on lister.cc:
```diff
@@ -75,3 +77,50 @@ future<> lister::rmdir(fs::path dir) {
+    auto ret = co_await _queue.pop_eventually();
+    if (!ret) {
+        // In case the caller tries to get() again
+        // after returning end of list.
```
Aborting the queue after it's exhausted will leave an exception for the *next* call to `get()`.
In that case, if `get()` is called yet-again (and I should probably add a test case for that),
`pop_eventually` should return an error since the queue is aborted.  (yes, we/I have made aborting the seastar::queue stateful so first thing `pop_eventually` does is to check if there's an exception set on the queue)

bhalevy replied:
Improved documentation in v4.

bhalevy replied:
And added a unit test for calling get() after the lister is done.

nyh replied:
I see now what I missed - you used a queue of optionals - so at the real end you pop a disengaged optional - here you don't pop anything (pop itself returns nothing) which means the disengaged optional was already popped last time. So ok.

bhalevy commented on test/boost/lister_test.cc:
```diff
@@ -108,3 +120,99 @@ SEASTAR_TEST_CASE(test_lister_abort) {
+        switch (*de->type) {
+        case directory_entry_type::regular: {
+            auto [it, inserted] = found_file_names.insert(de->name);
+            assert(inserted);
```
You're absolutely right.  My bad.  I had `generate_random_content` in mind.

bhalevy replied:
Fixed in v4.

bhalevy commented:
In v4:
- renamed variable for open_checked_directory: `f` -> `dir`
- renamed `_done` -> `_opt_done_fut`
- Improved documentation for `get()` and `close()`
- Set a default value for abort exception (to `broken_pipe_exception`)
- Test:
  - `s/assert(inserted)/BOOST_REQUIRE(inserted)/`
  - Added test case for `test_directory_lister_extra_get`

xemul commented:
@bhalevy , this breaks various debug-mode tests. E.g. the locator_test:
```
ERROR 2022-02-10 09:06:45,766 [shard 0] lister - directory_lister not closed when destroyed, at: 0x10229bd0 0x16999aed 0x169997ad 0x1699bd93 0x1699cc22 0x15ca1b2b 0x149a9cab 0x11c01495 0x10539f6d 0x10539d48 0x15d18b9e 0x15d1f216 0x15d23ebc 0x15d219f1 0x15b2db0c 0x15b2b677 0x15b2e2b7 0x102bf8d7 /lib64/libc.so.6+0x27b74 0x101ddcad
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void>::then_impl_nrvo<seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void> >(seastar::noncopyable_function<seastar::future<void> ()>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> ()>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::parallel_for_each_state
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void>::then_impl_nrvo<seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void> >(seastar::noncopyable_function<seastar::future<void> ()>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> ()>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::(anonymous namespace)::thread_wake_task
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void>::then_impl_nrvo<seastar::noncopyable_function<seastar::future<void> ()>, seastar::future<void> >(seastar::noncopyable_function<seastar::future<void> ()>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> ()>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)> >(seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::noncopyable_function<seastar::future<void> (seastar::future<void>&&)>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
Aborting on shard 0.
Backtrace:
  0x10229bd0
  0x15e8b48d
  0x15e8b08c
  0x15cf9f14
  0x15d25a6e
  0x15dfac1c
  0x15dfae6d
  0x15dfac9a
  0x7f5017561a1f
  /lib64/libc.so.6+0x3d2a1
  /lib64/libc.so.6+0x268a3
  0x15ca1c0b
  0x149a9cab
  0x11c01495
  0x10539f6d
  0x10539d48
  0x15d18b9e
  0x15d1f216
  0x15d23ebc
  0x15d219f1
  0x15b2db0c
  0x15b2b677
  0x15b2e2b7
  0x102bf8d7
  /lib64/libc.so.6+0x27b74
  0x101ddcad
```
I dequeued this PR

bhalevy commented:
In v5:
- directory_lister: get: auto-close on eol or error
  - make sure the lister is closed when get returns a disengaged entry or an error.

xemul commented on lister.cc:
```diff
@@ -101,11 +101,19 @@ future<std::optional<directory_entry>> directory_lister::get() {
+    }
+    co_await close();
+    if (ex) {
+        std::rethrow_exception(std::move(ex));
```
Should the be merged into patch 1 to make it git-bisect-testable-in-debug-mode?

bhalevy replied:
Sure

bhalevy replied:
Squashed in v6

xemul commented on lister.cc:
```diff
@@ -101,11 +101,19 @@ future<std::optional<directory_entry>> directory_lister::get() {
+    if (ex) {
+        std::rethrow_exception(std::move(ex));
     }
     co_return ret;
```
This is actually `co_return std::nullopt`, so the `std::optional<directory_entry> ret;` can be removed.

bhalevy replied:
Fixed in v6, plus a minor change:
```
        co_return coroutine::exception(std::move(ex));
```

instead of `std::rethrow_exception`

bhalevy commented:
In v6:
- co_return std::nullopt in the get() end-of-list case
  - also, co_return coroutine::exception instead of std::rethrow_exception
  - squashed into "lister: add directory_lister"

bhalevy commented:
@xemul / @nyh pin re-review

bhalevy commented:
@xemul / @nyh ping.  Please re-review
